In [15]:
import pandas as pd
import numpy as np

In [16]:
df = pd.read_csv('/content/adult.csv')

In [17]:
df.head()

,age,workclass,fnlwgt,education,education.num,marital.status,occupation,relationship,race,sex,capital.gain,capital.loss,hours.per.week,native.country,income
0,90,?,77053,HS-grad,9,Widowed,?,Not-in-family,White,Female,0,4356,40,United-States,<=50K
1,82,Private,132870,HS-grad,9,Widowed,Exec-managerial,Not-in-family,White,Female,0,4356,18,United-States,<=50K
2,66,?,186061,Some-college,10,Widowed,?,Unmarried,Black,Female,0,4356,40,United-States,<=50K
3,54,Private,140359,7th-8th,4,Divorced,Machine-op-inspct,Unmarried,White,Female,0,3900,40,United-States,<=50K
4,41,Private,264663,Some-college,10,Separated,Prof-specialty,Own-child,White,Female,0,3900,40,United-States,<=50K


In [18]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       32561 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education.num   32561 non-null  int64 
 5   marital.status  32561 non-null  object
 6   occupation      32561 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital.gain    32561 non-null  int64 
 11  capital.loss    32561 non-null  int64 
 12  hours.per.week  32561 non-null  int64 
 13  native.country  32561 non-null  object
 14  income          32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


In [19]:
df.shape

(32561, 15)

### Manual Section: Handling Missing Values

This section focuses on identifying and treating missing values in the dataset to ensure data quality for subsequent analysis.

# **Task - 1**

In [37]:
df.isnull().sum()

,0
age,0
workclass,1836
fnlwgt,0
education,0
education.num,0
marital.status,0
occupation,1843
relationship,0
race,0
sex,0


In [36]:
df.replace('?', np.nan, inplace=True)

### Verifying the replacement of missing value indicators

In [34]:
df.isnull().sum()

,0
age,0
workclass,1836
fnlwgt,0
education,0
education.num,0
marital.status,0
occupation,1843
relationship,0
race,0
sex,0


### **Task - 2**

In [39]:

# Calculate missing values count
missing_count = df.isnull().sum()

In [40]:
# Calculate missing values percentage
missing_percentage = (df.isnull().sum() / len(df)) * 100

In [45]:
missing_info = pd.DataFrame({
    'Missing Count': missing_count,
    'Missing Percentage': missing_percentage
})

#Filter for columns with missing values and sort by count in descending order
missing_info = missing_info[missing_info['Missing Count'] > 0].sort_values(by='Missing Count', ascending=False)

In [46]:
print("Missing values per column (count and percentage, sorted descending):")
display(missing_info)

Missing values per column (count and percentage, sorted descending):


,Missing Count,Missing Percentage
occupation,1843,5.660146
workclass,1836,5.638647
native.country,583,1.790486


### **Task - 3**

In [51]:
# Impute 'workclass' and 'occupation' with 'Unknown'
df['workclass'] = df['workclass'].fillna('Unknown')
df['occupation'] = df['occupation'].fillna('Unknown')

In [53]:
# Impute 'native.country' with 'Unknown'
df['native.country'] = df['native.country'].fillna('Unknown')

### Justification for Missing Value Imputation

For `workclass`, `occupation`, and `native.country`, missing values were imputed with 'Unknown' to retain data rows and treat missingness as a distinct, informative category. This prevents data loss and avoids assumptions about the true missing values.

In [52]:
# Impute 'workclass' and 'occupation' with 'Unknown'
df['workclass'] = df['workclass'].fillna('Unknown')
df['occupation'] = df['occupation'].fillna('Unknown')

In [50]:
# Verify that there are no more missing values in the affected columns
print("Missing values after imputation:")
display(df[['workclass', 'occupation', 'native.country']].isnull().sum())

Missing values after imputation:


,0
workclass,0
occupation,0
native.country,0


### **Task - 4**

In [66]:
print(f"We found {df.duplicated().sum()} rows that are exact duplicates.")

# Check for duplicates ignoring the 'income'
df_attributes = df.drop(columns=['income'])
num_duplicates_ignoring_target = df_attributes.duplicated().sum()
print(f"We also found {num_duplicates_ignoring_target} rows where people have the same information (like age, job, education), but diff income")

if num_duplicates_ignoring_target > 0:
    print("\nWhat does this mean? If people have the same background but different incomes, it could be a data error, or it might mean that 'income' is affected by things we don't have data for yet. For now, we'll just note them. Later, we might decide if we need to remove or change them, depending on what we want to find out from the data.")
else:
    print("\nNo duplicate rows found when we ignore the 'income' column.")

We found 24 rows that are exact duplicates.
We also found 25 rows where people have the same information (like age, job, education), but diff income

What does this mean? If people have the same background but different incomes, it could be a data error, or it might mean that 'income' is affected by things we don't have data for yet. For now, we'll just note them. Later, we might decide if we need to remove or change them, depending on what we want to find out from the data.


### **Task 5**

In [ ]:
df['education'].unique()

### Fixing 'education' column

In [56]:
print("Unique values in 'education' before stripping whitespace:")
print(df['education'].unique())

df['education'] = df['education'].str.strip()

print("\nUnique values in 'education' after stripping whitespace:")
print(df['education'].unique())

Unique values in 'education' before stripping whitespace:
['HS-grad' 'Some-college' '7th-8th' '10th' 'Doctorate' 'Prof-school'
 'Bachelors' 'Masters' '11th' 'Assoc-acdm' 'Assoc-voc' '1st-4th' '5th-6th'
 '12th' '9th' 'Preschool']

Unique values in 'education' after stripping whitespace:
['HS-grad' 'Some-college' '7th-8th' '10th' 'Doctorate' 'Prof-school'
 'Bachelors' 'Masters' '11th' 'Assoc-acdm' 'Assoc-voc' '1st-4th' '5th-6th'
 '12th' '9th' 'Preschool']


### Fixing 'marital-status' column

In [57]:
print("Unique values in 'marital-status' before stripping whitespace:")
print(df['marital.status'].unique())

df['marital.status'] = df['marital.status'].str.strip()

print("\nUnique values in 'marital-status' after stripping whitespace:")
print(df['marital.status'].unique())

Unique values in 'marital-status' before stripping whitespace:
['Widowed' 'Divorced' 'Separated' 'Never-married' 'Married-civ-spouse'
 'Married-spouse-absent' 'Married-AF-spouse']

Unique values in 'marital-status' after stripping whitespace:
['Widowed' 'Divorced' 'Separated' 'Never-married' 'Married-civ-spouse'
 'Married-spouse-absent' 'Married-AF-spouse']


### Fixing 'native-country' column

In [58]:
print("Unique values in 'native.country' before stripping whitespace:")
print(df['native.country'].unique())

df['native.country'] = df['native.country'].str.strip()

print("\nUnique values in 'native.country' after stripping whitespace:")
print(df['native.country'].unique())

Unique values in 'native.country' before stripping whitespace:
['United-States' 'Unknown' 'Mexico' 'Greece' 'Vietnam' 'China' 'Taiwan'
 'India' 'Philippines' 'Trinadad&Tobago' 'Canada' 'South'
 'Holand-Netherlands' 'Puerto-Rico' 'Poland' 'Iran' 'England' 'Germany'
 'Italy' 'Japan' 'Hong' 'Honduras' 'Cuba' 'Ireland' 'Cambodia' 'Peru'
 'Nicaragua' 'Dominican-Republic' 'Haiti' 'El-Salvador' 'Hungary'
 'Columbia' 'Guatemala' 'Jamaica' 'Ecuador' 'France' 'Yugoslavia'
 'Scotland' 'Portugal' 'Laos' 'Thailand' 'Outlying-US(Guam-USVI-etc)']

Unique values in 'native.country' after stripping whitespace:
['United-States' 'Unknown' 'Mexico' 'Greece' 'Vietnam' 'China' 'Taiwan'
 'India' 'Philippines' 'Trinadad&Tobago' 'Canada' 'South'
 'Holand-Netherlands' 'Puerto-Rico' 'Poland' 'Iran' 'England' 'Germany'
 'Italy' 'Japan' 'Hong' 'Honduras' 'Cuba' 'Ireland' 'Cambodia' 'Peru'
 'Nicaragua' 'Dominican-Republic' 'Haiti' 'El-Salvador' 'Hungary'
 'Columbia' 'Guatemala' 'Jamaica' 'Ecuador' 'France' 'Yugosla

### Brief Summary of Task 5: Fixing Inconsistent Categorical Data

For 'education', 'marital-status', and 'native-country', no leading or trailing whitespace inconsistencies were found after applying the `.str.strip()` function. This indicates these columns were already clean regarding this specific type of inconsistency.

### **Task - 6**

In [59]:
df.describe()

,age,fnlwgt,education.num,capital.gain,capital.loss,hours.per.week
count,32561.000000,3.256100e+04,32561.000000,32561.000000,32561.000000,32561.000000
mean,38.581647,1.897784e+05,10.080679,1077.648844,87.303830,40.437456
std,13.640433,1.055500e+05,2.572720,7385.292085,402.960219,12.347429
min,17.000000,1.228500e+04,1.000000,0.000000,0.000000,1.000000
25%,28.000000,1.178270e+05,9.000000,0.000000,0.000000,40.000000
50%,37.000000,1.783560e+05,10.000000,0.000000,0.000000,40.000000
75%,48.000000,2.370510e+05,12.000000,0.000000,0.000000,45.000000
max,90.000000,1.484705e+06,16.000000,99999.000000,4356.000000,99.000000


In [60]:
df.describe(include='object')

,workclass,education,marital.status,occupation,relationship,race,sex,native.country,income
count,32561,32561,32561,32561,32561,32561,32561,32561,32561
unique,9,16,7,15,6,5,2,42,2
top,Private,HS-grad,Married-civ-spouse,Prof-specialty,Husband,White,Male,United-States,<=50K
freq,22696,10501,14976,4140,13193,27816,21790,29170,24720


### Mean and Median Analysis

For both 'age' and 'hours.per.week', the mean and median values are very close. This indicates a relatively symmetrical distribution for both variables, suggesting that there are no extreme outliers significantly skewing the data.

### **Task - 7**

In [61]:
# (a) Most common occupation
print("Most common occupation:")
display(df['occupation'].value_counts().head(1))

Most common occupation:


,count
occupation,
Prof-specialty,4140


In [62]:
# (b) Percentage distribution of sex
print("\nPercentage distribution of sex:")
display(df['sex'].value_counts(normalize=True).mul(100).round(2).astype(str) + '%')


Percentage distribution of sex:


,proportion
sex,
Male,66.92%
Female,33.08%


In [63]:
# (c) Percentage distribution of income label
print("\nPercentage distribution of income:")
income_distribution = df['income'].value_counts(normalize=True).mul(100).round(2)
display(income_distribution.astype(str) + '%')


Percentage distribution of income:


,proportion
income,
<=50K,75.92%
>50K,24.08%


### Income Class Balance Commentary

The income classes are imbalanced. The '<=50K' category constitutes a much larger percentage (approximately 75.92 %) compared to '>50K' (approximately 24.08%). This imbalance should be considered in any predictive modeling tasks.

### **Task - 8**

In [65]:
education_consistency = df.groupby('education')['education.num'].unique()

print("Consistency check for 'education' and 'education.num':")
display(education_consistency)

# Check for inconsistencies (where an education level maps to more than one education.num)
inconsistencies = education_consistency[education_consistency.apply(len) > 1]

if not inconsistencies.empty:
    print("\n--- Inconsistencies Found ---")
    print("The following education levels map to multiple 'education.num' values:")
    display(inconsistencies)
else:
    print("\n--- No Inconsistencies Found ---")
    print("Each 'education' level consistently maps to a single 'education.num' value.")

Consistency check for 'education' and 'education.num':


,education.num
education,
10th,[6]
11th,[7]
12th,[8]
1st-4th,[2]
5th-6th,[3]
7th-8th,[4]
9th,[5]
Assoc-acdm,[12]
Assoc-voc,[11]



--- No Inconsistencies Found ---
Each 'education' level consistently maps to a single 'education.num' value.


In [ ]:
import sqlite3

conn = sqlite3.connect(':memory:')
df.to_sql('adult_income', conn, if_exists='replace', index=False)



In [ ]:

sql_query = "SELECT * FROM adult_income WHERE age > 30"
query_df = pd.read_sql_query(sql_query, conn)

print(f"Original DataFrame shape: {df.shape}")
print(f"Query result DataFrame shape (age > 30): {query_df.shape}")

# Display the first few rows of the queried DataFrame
print("\nFirst 5 rows of the queried DataFrame (age > 30):")
display(query_df.head())

conn.close()
print("\nSQLite connection closed.")

### **Task - 10**

### Data Profiling Summary Report

This report summarizes the dataset's structure, quality issues, key observations, and cleaning steps.

#### 1. Dataset Overview
*   **Size:** `adult.csv` has 32561 rows and 15 columns. It contains demographic and employment data, primarily for income prediction.

#### 2. Data Quality Issues Found
*   **Missing Values:** '?' was found in `occupation` (5.66%), `workclass` (5.64%), and `native.country` (1.79%).
*   **Duplicates:** 24 exact duplicate rows were found. 25 duplicates were found when ignoring the 'income' column, suggesting potential data entry variations.
*   **Inconsistent Categorical Data:** No leading/trailing whitespace issues were found in 'education', 'marital-status', or 'native-country' after checking.

#### 3. Key Observations
*   **Age & Hours per Week:** Both 'age' (mean: 39, median: 37) and 'hours.per.week' (mean: 40, median: 40) show symmetrical distributions with similar mean and median values.
*   **Most Common Occupation:** 'Prof-specialty'.
*   **Sex Distribution:** The dataset is predominantly 'Male' (67%).
*   **Income Imbalance:** Income classes are imbalanced: ~76% are '<=50K' and ~24% are '>50K'.

#### 4. Cleaning Decisions Made
*   **Missing Values:** '?' in `workclass`, `occupation`, and `native.country` were imputed with 'Unknown' to retain rows and treat missingness as a category.
*   **Whitespace:** Whitespace was stripped from 'education', 'marital-status', and 'native-country', but no inconsistencies were detected.